internal/pkg/agent/cmd/container.go

```diff
@@ -988,3 +991,129 @@ func isContainer(detail component.PlatformDetail) component.PlatformDetail {
+	}
+
+	reader, err := store.Load()
+	if err != nil {
+		return false, fmt.Errorf("failed to load from disk store: %w", err)
+	}
+
+	cfg, err := config.NewConfigFrom(reader)
+	if err != nil {
+		return false, fmt.Errorf("failed to read from disk store: %w", err)
+	}
+
+	storedConfig, err := configuration.NewFromConfig(cfg)
+	if err != nil {
+		return false, fmt.Errorf("failed to read from disk store: %w", err)
+	}
+
+	storedFleetHosts := storedConfig.Fleet.Client.GetHosts()
+	if len(storedFleetHosts) == 0 || !slices.Contains(storedFleetHosts, setupCfg.Fleet.URL) {
+		// fleet url changed thus elastic-agent should enroll
+		return true, nil
+	}
+
+	// here we don't return true if there is no previously saved enrollment token hash
+	// as an older version of elastic-agent won't have one.
+	if len(storedConfig.Fleet.EnrollmentTokenHash) > 0 {
+		// we have previously saved an enrollment token hash
+		enrollmentHashBytes, err := base64.StdEncoding.DecodeString(storedConfig.Fleet.EnrollmentTokenHash)
+		if err != nil {
+			return false, fmt.Errorf("failed to decode hash: %w", err)
+		}
+
+		err = crypto.ComparePBKDF2HashAndPassword(enrollmentHashBytes, []byte(setupCfg.Fleet.EnrollmentToken))
+		switch {
+		case errors.Is(err, crypto.ErrMismatchedHashAndPassword):
+			// the enrollment token hash does not match thus elastic-agent should enroll
+			return true, nil
+		case err != nil:
+			return false, fmt.Errorf("failed to compare hash: %w", err)
+		}
+	}
+
+	// Although fleet url and enrollment token are the same, we must also validate the api token.
+	// The agent might have been unenrolled from kibana which doesn't require for fleet url and
+	// enrollment token need to be changed
+	log, err := logger.New("fleet_client", false)
+	if err != nil {
+		return false, fmt.Errorf("failed to create logger: %w", err)
+	}
+	fc, err := fleetclient.NewAuthWithConfig(log, storedConfig.Fleet.AccessAPIKey, storedConfig.Fleet.Client)
+	if err != nil {
+		return false, fmt.Errorf("failed to create fleet client: %w", err)
+	}
+
+	// exploit the ACK request of fleet with **empty events**
+	// to validate the api token of the elastic-agent
+	ackRequest := &fleetapi.AckRequest{
+		Events: nil,
+	}
+	ackCMD := fleetapi.NewAckCmd(&agentInfo{storedConfig.Fleet.Info.ID}, fc)
+	_, err = ackCMD.Execute(ctx, ackRequest)
+	switch {
+	case errors.Is(err, fleetclient.ErrInvalidAPIKey):
+		// this elastic-agent has not valid api token thus it should enroll
+		return true, nil
+	case err != nil:
+		return false, fmt.Errorf("failed to validate api token: %w", err)
+	}
+
+	if len(storedConfig.Fleet.EnrollmentTokenHash) == 0 {
+		// previous EnrollmentTokenHash was empty, but we just validated
+		// that this elastic-agent is still eligible to authenticate with fleet.
+		// Update the config with the validated EnrollmentTokenHash.
+		enrollmentHashBytes, err := crypto.GeneratePBKDF2FromPassword([]byte(setupCfg.Fleet.EnrollmentToken))
```

Comment:
What happens if they have updated the container to remove the EnrollmentToken but are using the same storage of an already enrolled container? Maybe this should check that an EnrollmentToken is set in the `setupCfg` before it just hashes it?

Reply:
if the enrollmentToken is empty and the one we have in our state isn't wouldn't that cause the two hashes to be different and then the agent to re-enroll? What am I missing 🤔

Reply:
I don't think your missing anything, just wanted to be sure that it would work.

Reply:
thanks for the comment, actually you made me think of another case where the fleet enrollment token from the setup can be empty and the agent should fallback to kibana to extract a new one. I updated the code to capture that case as well. Feel free to unresolve if you spot anything else 🙂